rfcs/0003-api-unification.md

```diff
@@ -0,0 +1,277 @@
+- Feature Name: API interface
+- Start Date: 2018-10-30)
+- RFC PR:
+- Salt Issue:
+
+# Summary
+[summary]: #summary
+
+Salt Module Interface (SMI) concept introduction for virtual modules.
+
+# Motivation
+[motivation]: #motivation
+
+Any salt module has couple of specific properties we dealing every day
+with:
+
+- Fixed set of functions
+- Known functions signatures
+- Known structure of return
+
+This is true but for virtual modules. The virtual module covering
+several "fixed" or "physical" modules and behaves like that would be
+one module. But differences between those physical modules on
+different platforms makes such virtual module a moving target and
+unpredictable.
+
+Virtual modules concept is missing crucial part in the design:
+interfaces. The interface should define how module looks like and what
+APIs can be called to it. Interface should move module that is
+_called_ differently on heterogeneous environments to a module that
+_reports_ differently on heterogeneous environments.
+
+
+# Design
+[design]: #detailed-design
+
+*DISCLAIMER*: The SMI is not that classic understanding of typical
+interface one may find in languages like Java. It is also not as same
+as Zope Interface package or Python Abstract Base Classes (ABC).
```

Comment:
I think we should further explore why we _shouldn't_ use Python ABCs. Part of my concern here is how application-specific this proposal is. I think we would be best served by being as "Pythonic" as possible.

Reply:
@cachedout The interface concept I am proposing is to dissect what should be for virtual modules (commonly available functions) from what is not everywhere possible.

We cannot use ABCs because first of all they stand for "abstract base **classes**". We do not have this by design, since all modules in Salt are functions at the module level. The ABC mechanism is pretty simple: as long as there is abstract method which is not overridden by an actual implementation, the `@abstractmethod` decorator will just trigger TypeError exception and block _abstract class_ from being instantiated with abstract methods in it.

But even if we would use classes, we still don't want that behaviour as well. We want to be very transparent and make sure that outside of the interface methods are perfectly usable, just they are not visible through the virtual access. That is that `mymodule.fancy_function` is accessible but `myvirtmodule.fancy_function` is not there, although this is call of the same module, just first is physical, the other is virtual. Because if someone needs something that is only possible on OSX or Solaris, we should not kill the function or force everyone to implement "Unsupported!" stub. But yes, some modules might be unsupported, in that case automatically added stub actually makes sense.